ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java

```diff
@@ -112,24 +115,33 @@ public void removeXAResourceRecoveryHelper(XAResourceRecoveryHelper xaResourceRe
                  * the target xaResourceRecoveryHelper and if so then we need to wait for second pass
                  * of the scanner to finish
                  */
-                if (isHelperInUse(xaResourceRecoveryHelper))
+                if (isHelperInUse(xaResourceRecoveryHelper)) {
                     waitForScanState(ScanStates.IDLE);
-                else if (getScanState().equals(ScanStates.BETWEEN_PASSES)) {
-					synchronized (this) { // Because we need to remove the _resources
-						XAResource[] xaResources = recoveryHelpersXAResource.get(xaResourceRecoveryHelper);
-						if (xaResources != null) {
-							for (XAResource xar : xaResources) {
-								xaRecoverySecondPass(xar);
-								_resources.remove(xar);
-							}
-						} else {
-							System.out.println("Was null");
-						}
-					}
-				}
+                } else {
+                    // synchronized(this) is needed as removing on the _resources; to avoid deadlock
+                    //   we need to preserve the order of acquiring locks - first 'this' then second 'scanStateLock'
+                    scanStateLock.unlock();
+                    synchronized (this) {
+                        scanStateLock.lock();
```

Comment:
It looks a little bit strange that you release the lock and then reacquire it. From what I can see though we do only need to check for the state BETWEEN_PASSES to do the removes but I wonder if we should check again for something like 

```
isHelperInUse(xaResourceRecoveryHelper)) {
                    waitForScanState(ScanStates.IDLE);
}
```

Under the second scanStateLock.lock() code?